### PR TITLE
Addressing various issues

### LINKED ISSUE #9: Enable semantic coloring of actor types that currently cannot be colored
Enable coloring of:

- [x] Skeletal meshes
- [x] Landscapes

### LINKED ISSUE #17: Add camera dimensions to CSV output

### LINKED ISSUE #24: Create an output file with semantic color - semantic class mapping when rendering semantics

### LINKED ISSUE #25: Explain linear and radial depth

### LINKED ISSUE #27: How to use optical flow data ?
I use this code to covert optical flow data.But it doesn't look right.
Here is  my python code:
```
import cv2
import torch
import torch.nn as nn
import numpy as np
def img2flow(img,scale):
    '''
    :param img: cv img bgr
    :param scale: optical scale
    :return: dx dy tensor
    '''
    h = img.shape[0]
    w = img.shape[1]
    img = cv2.cvtColor(img,cv2.COLOR_BGR2HSV) #h is angle ,s is intensity
    ang = img[:,:,0] #* 2 *np.pi / 180
    mag = img[:,:,1]  / scale
    dx ,dy = cv2.polarToCart(mag,ang,angleInDegrees=True)
    dx = w * dx
    dy = h *dy
    dx = torch.Tensor(dx).unsqueeze(0).unsqueeze(0)
    dy = torch.Tensor(dy).unsqueeze(0).unsqueeze(0)
    #print(dx.shape)
    print(dx)
    print(dx.max())




    return dx ,dy
def read_img2tensor(path):
    img = cv2.imread(path)
    img = cv2.cvtColor(img,cv2.COLOR_BGR2RGB)
    img = img.transpose(2, 0, 1)
    tensor = torch.Tensor(img).float().unsqueeze(0)

    #1*3*h*w

    return tensor
def vis_tensor(winname,tensor):
    #tensor 1 3 h w
    mat = tensor.squeeze().detach().numpy()
    mat = np.uint8(mat)  # float32-->uint8
    mat = mat.transpose(1, 2, 0)  # mat_shape: (982, 814，3)
    mat = cv2.cvtColor(mat,cv2.COLOR_BGR2RGB)
    #cv2.imshow(winname,mat)
    cv2.imwrite(r'E:\data\EasySynth_test2/'+winname+'.jpg',mat)
    return 0

def warp(x, flo):
    """
    warp an image/tensor (im2) back to im1, according to the optical flow
    x: [B, C, H, W] (im2)
    flo: [B, 2, H, W] flow
    """
    B, C, H, W = x.size()
    # mesh grid
    xx = torch.arange(0, W).view(1, -1).repeat(H, 1)
    yy = torch.arange(0, H).view(-1, 1).repeat(1, W)
    xx = xx.view(1, 1, H, W).repeat(B, 1, 1, 1)
    yy = yy.view(1, 1, H, W).repeat(B, 1, 1, 1)
    grid = torch.cat((xx, yy), 1).float()

    # if x.is_cuda:
    #     grid = grid.cuda()
    vgrid = grid + flo

    # scale grid to [-1,1]
    vgrid[:, 0, :, :] = 2.0 * vgrid[:, 0, :, :].clone() / max(W - 1, 1) - 1.0
    vgrid[:, 1, :, :] = 2.0 * vgrid[:, 1, :, :].clone() / max(H - 1, 1) - 1.0

    vgrid = vgrid.permute(0, 2, 3, 1)
    output = nn.functional.grid_sample(x, vgrid,mode="bilinear",align_corners=False)
    mask = torch.autograd.Variable(torch.ones(x.size()))
    mask = nn.functional.grid_sample(mask, vgrid)

    # if W==128:
    # np.save('mask.npy', mask.cpu().data.numpy())
    # np.save('warp.npy', output.cpu().data.numpy())

    mask[mask < 0.9999] = 0
    mask[mask > 0] = 1

    return output * mask


if __name__ == '__main__':
    #'exr'
    path0 = r'E:\data\EasySynth_test\ColorImage/'+'testSeq.0008.jpeg'
    path1 = r'E:\data\EasySynth_test\ColorImage/' + 'testSeq.0007.jpeg'
    #flow_path = r'E:\data\EasySynth_test\OpticalFlowImage_scale2/' + 'testSeq.0008.exr'
    flow_path = r'E:\data\EasySynth_test\OpticalFlowImage_scale1/' + 'testSeq.0008.exr'
    flow_img = cv2.imread(flow_path, cv2.IMREAD_ANYCOLOR | cv2.IMREAD_ANYDEPTH)
    #flow_img = cv2.imread(flow_path)
    print(flow_img.max())
    dx , dy = img2flow(flow_img,scale=1)
    flow = torch.cat((dx,dy),dim=1)

    #test
    #read img0 and img1
    img0 = read_img2tensor(path0)
    img1 = read_img2tensor(path1)
    img1_0 =warp(img1,flow)

    vis_tensor('0',img0)
    vis_tensor('0warp',img1_0)
```
testSeq.0007.jpeg:

testSeq 0007
testSeq.0008.jpeg:
testSeq 0008
And after warp 7 to 8 with testSeq.0008.exr looks like this:
0warp
I also tried other ways of warp. such from 0008.jpeg warp to 0007.jpeg. 
Is there something wrong?

Besides ,when I use .jpeg format to get optical flow data with replace code cv2.imread,the value will be far out of range.And in this case scale doesn't seem to be used correctly in my code.

### LINKED ISSUE #29: Visible aliasing in rendered images
Rendered images have visible aliasing artifacts, screenshots attached. 

UE4, Ubuntu Linux 20.04 LTS

Screenshot from 2022-05-26 11-33-26

Screenshot from 2022-05-26 11-33-12

Screenshot from 2022-05-26 11-33-04

Screenshot from 2022-05-26 12-07-45
.

### LINKED ISSUE #32: Empty PNG output with PostProcess Material used after Tonemapping
**Describe the bug**

png output are empty for all of types of renders. 

**To Reproduce**

1. Check main branch. 
2. Create new unreal scene. 
3. Add any object (sphere, cube)
4. Add new sequencer with camera
5. Use the plugin as stated in manual
6. The png output is empty -> probably transparency is 100% (not sure)
(this is example image)
Run1_0_0000

**Configuration (if applicable):**
 - OS: Windows 10
 - GPU: Nvidia 3070
 - Unreal: 4.27
**Additional context**
The problem is connected with the tonemapper. When I change the mode of PPMaterial from "After Tonemapper" to "Before Tonemapper", the output exists. The problem is the same for the setting "Replace Tonemapper". Have you encountered the issue by yourself?

### LINKED ISSUE #33: Creating a stereo camera ground truth rig is difficult with camera's within blueprints or attached to other actors
Creating a stereo camera ground truth rig is difficult with cameras within blueprints or cine camera actors attached to other actors (see image below). In my workflow, I have attempted to use take recorder to record a vehicle blueprint with attached camera component and as well as using a level sequence and attaching a camera to a primary camera actor which I navigate in the level sequence. For the attached camera or camera within the blueprint, the camera poses can not be exported (EasySynt. The primary camera's camera pose in the example below can be exported.

image

image

### LINKED ISSUE #35: Enable users to opt for constant camera movement (without speedup at the start)
Either expand the README or add a button, whichever offers the best solution.

### PR DESCRIPTION
**Pull request by NikolaJov96**
This PR introduces multi-camera rigs and various fixes and improvements described in closed issues.

Closes #9 - Painting skeletal meshes and landscapes enabled
Closes #17 - Included with camera rig changes
Closes #24 - Added import and export option for semantic labels
Closes #25 - Added to the README
Closes #27 - Added explanation to the README, changing image size not fixed yet
Closes #29 - Added explanation to the README
Closes #32 - Fixed post-process materials to produce valid alpha
Closes #33 - Enabled multi-camera rigs, no fix for the Take Recorder
Closes #35 - Added advice to the README

### CODE REVIEW

**Review by fpanjevic — CHANGES_REQUESTED** [review]

File: Source/EasySynth/Public/CameraRig/CameraRigRosInterface.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+	/** Handles exporting camera rig into a ROS JSON file */
+	bool ExportCameraRig(
+		const FString& OutputDir,
+		TArray<UCameraComponent*> RigCameras,
+		const FIntPoint& SensorSize);
+
+private:
+	/** Adds lines describing a single camera to the output array */
+	void AddCamera(
```
[the remark this answers is not in the document]
Comment (NikolaJov96) [reply Source/EasySynth/Public/CameraRig/CameraRigRosInterface.h:30]: There is, but it seems to be cumbersome for this simple use case, I'll give it a try

[the remark this answers is not in the document]
Comment (NikolaJov96) [reply Source/EasySynth/Public/CameraRig/CameraRigRosInterface.h:88]: `FJsonObjectConverter` is now used for JSON import and export, the code should be a bit prettier

File: README.md [hunk 2]
```diff
@@ -212,42 +214,36 @@ for i, pose in poses_df.iterrows():
...
-      "rotation": [
-         [1.000000, -0.000000, 0.000000],
-         [0.000000, -1.000000, -0.000000],
-         [0.000000, 0.000000, -1.000000]
-      ],
-      "translation": [0.300000, -0.000000, 0.000000],
-      "sensor_size": [1920.000000, 1080.000000]
-    }
-  }
+	"cameras":
```
Comment (fpanjevic) [inline README.md:227]: <3

**Review by fpanjevic — APPROVED** [review]

**Review by fpanjevic — APPROVED** [review]

**Comment by fpanjevic** [comment]
Ping @rankela

File: README.md [hunk 3]
```diff
@@ -184,6 +210,42 @@ for i, pose in poses_df.iterrows():
...
+Following is a ROS JSON file example for a rig with two parallel cameras with the FOV of 90 degrees, facing forward. The difference can be found in the sign of the translation vector X-axis.
+
+``` JSON
+{
+	"cameras":
+	{
+		"c0":
+		{
+			"intrinsics": [ 960, 0, 960, 0, 960, 540, 0, 0, 0 ],
+			"coord_sys": "RDF",
```
Comment (rankela) [inline README.md:232]: We can omit this field since it is unused

File: Source/EasySynth/Private/CameraRig/CameraRigRosInterface.cpp [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+		CameraData.CameraName = CameraName;
+
+		// Get the sensor size
+		if (RosJsonCamera.sensor_size.Num() != 2)
+		{
+			const FText ErrorMessage = LOCTEXT("InvalidSensorError", "Expected 2 values for the sensor size");
+			DisplayError(ErrorMessage);
+			return FReply::Handled();
+		}
+		CameraData.SensorSize.X = RosJsonCamera.sensor_size[0];
```
Comment (rankela) [inline Source/EasySynth/Private/CameraRig/CameraRigRosInterface.cpp:93]: Can we use sensor size to modify Camera aspect ratio? I noticed I have to do it manually for all cameras after import

Comment (rankela) [reply Source/EasySynth/Private/CameraRig/CameraRigRosInterface.cpp:93]: I mean Sensor Width/Height params, if aspect ratio is not correct rendered images are wrong size
image

File: Source/EasySynth/Private/CameraRig/YAML/CameraRigYamlInterface.cpp [hunk 5]
```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2022 YDrive Inc. All rights reserved.
```
Comment (rankela) [inline Source/EasySynth/Private/CameraRig/YAML/CameraRigYamlInterface.cpp:1]: I don't think anything needed for YAML is needed anymore. If you wish to keep it to support people using OpenCV for calibration that's fair, but it might be worthwhile to unify JSON and YML interfaces a bit more for maintenance reasons

File: Source/EasySynth/Public/CoordinateSystemConverter.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2022 YDrive Inc. All rights reserved.
```
Comment (rankela) [inline Source/EasySynth/Public/CoordinateSystemConverter.h:1]: Isn't the plan to remove this file altogether?

**Review by rankela — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (NikolaJov96) [reply README.md:232]: I thought it might be useful for clarity, will remove it

File: Source/EasySynth/Private/CameraRig/CameraRigRosInterface.cpp (hunk 4 above)
Comment (NikolaJov96) [reply Source/EasySynth/Private/CameraRig/CameraRigRosInterface.cpp:93]: Good catch, don't know how it didn't surface earlier.
Actually, the camera sensor width should be adjusted before rendering to match the requested output image aspect ratio.

File: Source/EasySynth/Private/CameraRig/YAML/CameraRigYamlInterface.cpp (hunk 5 above)
Comment (NikolaJov96) [reply Source/EasySynth/Private/CameraRig/YAML/CameraRigYamlInterface.cpp:1]: Will remove it, just didn't want to delete the code before we were sure we won't use it

File: Source/EasySynth/Public/CoordinateSystemConverter.h (hunk 6 above)
Comment (NikolaJov96) [reply Source/EasySynth/Public/CoordinateSystemConverter.h:1]: It is removed in the `coordinate_systems` branch, will merge it here when we are sure it works correctly

File: Source/EasySynth/Private/CameraRig/CameraRigRosInterface.cpp [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+
+	// Add intrinsics
+	const double FocalLength = SensorSize.X / UKismetMathLibrary::DegTan(Camera->FieldOfView / 2.0f) / 2.0f;
+	const double PrincipalPointX = SensorSize.X / 2.0f;
+	const double PrincipalPointY = SensorSize.Y / 2.0f;
+	RosJsonCamera.intrinsics.Init(0, 9);
+	RosJsonCamera.intrinsics[0] = FocalLength;
+	RosJsonCamera.intrinsics[2] = PrincipalPointX;
+	RosJsonCamera.intrinsics[4] = FocalLength;
+	RosJsonCamera.intrinsics[5] = PrincipalPointY;
```
Comment (rankela) [inline Source/EasySynth/Private/CameraRig/CameraRigRosInterface.cpp:243]: RosJsonCamera.intrinsics[8] = 1.0f

File: README.md [hunk 8]
```diff
@@ -161,31 +161,24 @@ Output is the `CameraPoses.csv` file, in which the first line contains column na
...
 
-The coordinate system for saving camera positions and rotation quaternions is a right-handed coordinate system. When looking through a camera with zero rotation in the target coordinate system:
-- X axis points to the right
-- Y axis points down
-- Z axis points straight away from the camera
-
-<img src="ReadmeContent/OutputCoordinateSystem.png" alt="Output coordinate system" width="250" style="margin:10px"/>
-
-Note that this differs from Unreal Engine, which internally uses the left-handed Z-up coordinate system.
+The coordinate system for saving camera positions and rotation quaternions is the same one used by Unreal Engine, a left-handed Z-up coordinate system. This system is pretty much only used by UE and will most likely require a conversion for your later use. Still, it seems to be the cleanest option, as exported values will match the numbers displayed inside the engine.
```
Comment (fpanjevic) [inline README.md:173]: > The coordinate system for saving camera positions and rotation quaternions is the same one used by Unreal Engine, a ***left-handed*** Z-up coordinate system.

Coordinates will ***likely require conversion*** to more common reference frames for typical computer vision applications. For more information we recommend [this Reddit post](https://www.reddit.com/r/gamedev/comments/7qh3sa/a_coordinate_system_chart_of_different_engines/).

File: README.md [hunk 9]
```diff
@@ -212,15 +205,18 @@ for i, pose in poses_df.iterrows():
...
 - `intrinsics` - Camera intrinsics matrix
-- `coord_sys` - Type of the coordinate system, `RDF` is required
 - `rotation` - Rotation quaternion relative to the rig origin
 - `translation` - Translation vector relative to the rig origin
 - `sensor_size` - Sensor width and height in pixels, used to calculate camera FOV
 
-Following is a ROS JSON file example for a rig with two parallel cameras with the FOV of 90 degrees, facing forward. The difference can be found in the sign of the translation vector X-axis.
+Rotation and translation values follow the UE coordinate system convention and are relative to the rig origin.
+
+Camera sensor sizes and aspect ratios will not be updated at this point, but they will be calculated according to the requested output image size when rendering starts.
```
Comment (fpanjevic) [inline README.md:217]:
```suggestion
Camera sensor sizes and aspect ratios will not be updated at this point, but they will be calculated according to the requested output image size when the rendering starts.
```

**Review by fpanjevic — CHANGES_REQUESTED** [review]
Much cleaner!

**Review by rankela — APPROVED** [review]
